File: src/api/workspace.ts
import { invokeWithResponseSchema } from "@/api/ipc";
import {
	EmptyWorkspaceResponseSchema,
	WorkspaceSchema,
	WorkspacesSchema,
} from "@/types/workspace";

/** Lists every reusable Workspace for navigation and Task composition. */
const listWorkspaces = () =>
	invokeWithResponseSchema("list_workspaces", WorkspacesSchema);

/** Registers a user-owned folder without transferring ownership to Theoria. */
const registerExternalWorkspace = (name: string, sourcePath: string) =>
	invokeWithResponseSchema("register_external_workspace", WorkspaceSchema, {
		request: { name, sourcePath },
	});

/** Creates an empty Workspace whose template files are owned by Theoria. */
const createManagedWorkspace = (name: string) =>
	invokeWithResponseSchema("create_managed_workspace", WorkspaceSchema, {
		request: { name },
	});

/** Changes whether a Workspace is pinned and returns its persisted state. */
const setWorkspacePin = (workspaceId: string, isPinned: boolean) =>
	invokeWithResponseSchema("set_workspace_pin", WorkspaceSchema, {
		request: { workspaceId, isPinned },
	});

/** Changes one persisted Workspace name after backend validation. */
const renameWorkspace = (workspaceId: string, name: string) =>
	invokeWithResponseSchema("rename_workspace", WorkspaceSchema, {
		request: { workspaceId, name },
	});

/** Removes a Workspace collection while requiring confirmation for managed files. */
const removeWorkspace = (workspaceId: string, managedFilesConfirmed: boolean) =>
	invokeWithResponseSchema("remove_workspace", EmptyWorkspaceResponseSchema, {
		request: { workspaceId, managedFilesConfirmed },
	});

export {
	createManagedWorkspace,
	listWorkspaces,
	registerExternalWorkspace,
	removeWorkspace,
	renameWorkspace,
	setWorkspacePin,
};

File: src-tauri/src/commands/workspace.rs
use crate::dto::workspace::{
    CreateManagedWorkspaceRequest, RegisterExternalWorkspaceRequest, RemoveWorkspaceRequest,
    RenameWorkspaceRequest, SetWorkspacePinRequest, WorkspaceResponse,
};
use crate::error::IpcError;
use crate::services::cleanup::WorkspaceCleanupService;
use crate::services::workspace::WorkspaceService;
use tauri::State;

/// Registers an external Workspace without modifying the selected directory.
#[tauri::command]
pub(crate) async fn register_external_workspace(
    request: RegisterExternalWorkspaceRequest,
    service: State<'_, WorkspaceService>,
) -> Result<WorkspaceResponse, IpcError> {
    service
        .register_external(request.name, request.source_path)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

/// Creates a managed Workspace template in application storage.
#[tauri::command]
pub(crate) async fn create_managed_workspace(
    request: CreateManagedWorkspaceRequest,
    service: State<'_, WorkspaceService>,
) -> Result<WorkspaceResponse, IpcError> {
    service
        .create_managed(request.name)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

/// Lists reusable Workspace inputs for navigation and selection.
#[tauri::command]
pub(crate) async fn list_workspaces(
    service: State<'_, WorkspaceService>,
) -> Result<Vec<WorkspaceResponse>, IpcError> {
    service
        .list()
        .await
        .map(|items| items.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}

/// Changes one Workspace name after applying the creation-time validation rules.
#[tauri::command]
pub(crate) async fn rename_workspace(
    request: RenameWorkspaceRequest,
    service: State<'_, WorkspaceService>,
) -> Result<WorkspaceResponse, IpcError> {
    service
        .rename(&request.workspace_id, request.name)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

/// Changes whether one Workspace is pinned to the top of the sidebar.
#[tauri::command]
pub(crate) async fn set_workspace_pin(
    request: SetWorkspacePinRequest,
    service: State<'_, WorkspaceService>,
) -> Result<WorkspaceResponse, IpcError> {
    service
        .set_pin(&request.workspace_id, request.is_pinned)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

/// Removes Workspace Tasks and mounts while protecting external source files.
#[tauri::command]
pub(crate) async fn remove_workspace(
    request: RemoveWorkspaceRequest,
    service: State<'_, WorkspaceCleanupService>,
) -> Result<(), IpcError> {
    service
        .remove_workspace(&request.workspace_id, request.managed_files_confirmed)
        .await
        .map_err(Into::into)
}
